### ISSUE TITLE
[FEATURE] Refactor the installer console UI and configuring process

### ISSUE THREAD

**Issue by weihanglo** [harvester/harvester#1422]
**Is your feature request related to a problem? Please describe.**

The current console installer UI has bad smell:
- Highly coupled with configuration workflow and validation process, which makes new feature hard to add.
- The current UI library lacks out-of-box widgets which cost efforts to implement all the widget from Harvester side.

We need to make sure any subsequent new config/UI feature request can be implemented easily.

**Describe the solution you'd like**

Decouple UI layout logic from other parts, as well as introduce new full-fledged console UI library if possible.

- Decouple UI logic from configuration workflow and validation process. YaST describes [a good architecture](https://yastgithubio.readthedocs.io/en/latest/architecture/) for this part.  
- Leverage a new UI library such as [tview](https://github.com/rivo/tview) or just YaST since they provide lots widgets out-of-box.

---

## Problems of v0.3.0 installer

- Lack of layout systems: `gocui` uses a naive canvas-like way to render stuff.
- Lack of out-of-box widgets. Speficially we need multi-select dropdown.
- No page management: We destroy old components imperatively.
- No auto-scroll widgets: the installer may be out of dimension and crashes.

### Pros

- Two layout systems: Flex and Grid. We don't need to layout widgets by hands (hope so).
- Out-of-box widgets: TextInput, TextView, DropDown, Form, Modal, etc.
- Rendering becomes More declarative less imperative.
- Focus-based widget controls: you can only interact with focused widgets.

### Cons

- Still no multi-select widget.
- The logic of auto-scroll and input-handler is highly coupled with widgets. We'll lose some benefits by customizing from the ground up.
- Cannot call `Focus` imperatively so it leads to customize our own widgets. As a result, it's not that easy to implement validation hooks when proceeding the workflow.

## Alternatives

- Leverage [`YaST`](https://yast.opensuse.org/), which provides "write once, run in both text and graphic UI".
- Leave old UI alone. First we need to know how many features are requested for installer UI.

**Comment by weihanglo** [comment]
Here are some ideas concerning the refactor

### A centeralised page navigation stack

- Won't need to handle `pop` actions (go to previous page). Write logic only for `push` actions (go to next).
- Register each page in the first place. When needs a navigation, just refers to the registered name.

### Hook functions on widgets

- Change from imperative style to hook (callback) style, e.g., `onClick` and `onSubmit`.
- (Optional) Reactive rendering. Widget updateds once bond value updates.


### Separate each UI component from domain logic

- Make validator as pure composable function that is easy to test
- UI components can compose multiple validator into one and call it in components' hook funcdtions

**Comment by votdev** [comment]
Here are my thoughts concerning the refactoring.

The installer needs to be improved because it currently lacks in UX and a professional look. It is the first thing that users/customers/decision makers see when they install (and test) Harvester. And we all know that first impressions count.

- There is the need of a welcome screen, including the project name (our cool retro ASCII logo) and a welcome message to the user.
- Each page should have a short disclaimer what is happening on the specific page.
- It should be possible to navigate forwards and backwards between the pages.
- The password page should mention the name of the specific user that this password is assigned to. A disclaimer should also mention that this can be used later on to access the machine via TTY or SSH.

Here are some suggestions of TUI frameworks that can be used. Unfortunately, there is not as much choice as with other languages like Python or Rust. Maybe it is not even necessary to use a TUI framework, but simply a wrapper built around the often used `dialog` CLI tool.

# Wrapper
- https://github.com/ScriptRock/dialog
- https://github.com/dorzheh/go-dialog

# TUI libraries
- https://github.com/VladimirMarkelov/clui
- https://pterm.sh/
- https://github.com/charmbracelet/bubbletea
- https://github.com/rivo/tview
- https://github.com/gdamore/tcell

**Comment by tserong** [comment]
I don't have any strong opinions about the various TUI frameworks, but just wanted to add that the installer needs to work via serial console, which irritatingly seems to be limited to 80 columns x 20-something (23? 24?) rows. The current implementation expects the terminal to be taller than that, and fails badly if it's not, requiring workarounds. This has come up several times in the past:

https://github.com/harvester/harvester/issues/2708
https://github.com/harvester/harvester/issues/3446
https://github.com/harvester/harvester/issues/1792

To put it another way: whatever solution we arrive at, needs to be forgiving of small / weird terminals :-)
